src/dev/stage_by_codeowner.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import simpleGit from 'simple-git';
+import fs from 'fs';
+import path from 'path';
+import { run } from '@kbn/dev-cli-runner';
+
+const git = simpleGit();
+
+const CODEOWNERS_PATH = path.join('.github', 'CODEOWNERS');
+
+// Function to get the list of changed files
+const getChangedFiles = async () => {
+  const status = await git.status();
+  return status.files.map((file) => file.path);
+};
+
+run(async ({ flags, log }) => {
+  const {
+    _: [owner],
+  } = flags;
+
+  if (!fs.existsSync(CODEOWNERS_PATH)) {
+    log.error('CODEOWNERS file not found!');
+    process.exit(1);
+  }
+
+  const changedFiles = await getChangedFiles();
+  const owners: Record<string, string[]> = {};
+
+  const getOwner = (file: string) => {
+    const codeowners = fs
+      .readFileSync(CODEOWNERS_PATH, 'utf-8')
+      .split('\n')
+      .filter((line) => line && !line.trim().startsWith('#'));
+
+    for (const line of codeowners) {
+      const [pattern, fileOwner] = line.split(/\s+/);
+      if (file.startsWith(pattern)) {
+        return fileOwner;
+      }
+    }
+
+    log.warning(`No owner found for ${file}`);
+
+    return null;
+  };
```

Comment:
I think this logic doesn't always find the correct code owner:
- AFAIU the `startsWith` would not properly match glob patterns in code owner definitions
- This will return the first match in the codeowners file, however the GH codeowners logic uses the last matching entry of that file

We have a package to check code ownership for files: https://github.com/elastic/kibana/blob/main/packages/kbn-code-owners/src/file_code_owner.ts
(when checking multiple files, you should run `getPathsWithOwnersReversed` once and store the lookup table, then pass it to your calls of `getCodeOwnersForFile`).

Reply:
FYI, I have a [PR](https://github.com/elastic/kibana/pull/204023) up with improvements for the `@kbn/code-owners` package.

Once that goes in, if you want to use the `@kbn/code-owners` package, it'd look something like this:

```suggestion
  const codeOwnersEntries = getCodeOwnersEntries();
  const getOwner = (file: string) => {
    const owners = getOwningTeamsForPath(file, codeOwnersEntries);
  
    if (owners.length === 0) {
      log.warning(`No owner found for ${file}`);
      return null;
    }
  
    return owners.join(',');
  };
```

+ imports
```ts
import { getOwningTeamsForPath, getCodeOwnersEntries } from '@kbn/code-owners';
```